contracts/contracts/token/OUSD.sol

```diff
@@ -466,25 +466,34 @@ contract OUSD is Initializable, InitializableERC20Detailed, Governable {
      * @dev Ensures internal account for rebasing and non-rebasing credits and
      *      supply is updated following deployment of frozen yield change.
      */
-    function _ensureRebasingMigration(address _account) internal {
-        if (nonRebasingCreditsPerToken[_account] == 0) {
-            if (_creditBalances[_account] == 0) {
-                // Since there is no existing balance, we can directly set to
-                // high resolution, and do not have to do any other bookkeeping
-                nonRebasingCreditsPerToken[_account] = 1e27;
-            } else {
-                // Migrate an existing account:
-
-                // Set fixed credits per token for this account
-                nonRebasingCreditsPerToken[_account] = _rebasingCreditsPerToken;
-                // Update non rebasing supply
-                nonRebasingSupply = nonRebasingSupply.add(balanceOf(_account));
-                // Update credit tallies
-                _rebasingCredits = _rebasingCredits.sub(
-                    _creditBalances[_account]
-                );
-            }
+    function _ensureMigrationToNonRebasing(address _account) internal {
+        if (nonRebasingCreditsPerToken[_account] != 0) {
+            return; // Account already is non-rebasing
+        }
+        if (_creditBalances[_account] == 0) {
+            // Since there is no existing balance, we can directly set to
+            // high resolution, and do not have to do any other bookkeeping
+            nonRebasingCreditsPerToken[_account] = 1e27;
+            return;
         }
+
+        // This does not change, but if it did, we would want to
+        // use the value before changes.
+        uint256 oldCredits = _creditBalances[_account];
+
+        // Atomicly update account information:
+        // It is important that balanceOf not be called inside updating
+        // account data, since it will give wrong answers if it does
+        // not have all an account's data in a consistent state.
+        //
+        // By setting a per account nonRebasingCreditsPerToken,
+        // this account will no longer follow with the global
+        // rebasing credits per token and will become non-rebasing.
+        nonRebasingCreditsPerToken[_account] = _rebasingCreditsPerToken;
+
+        // Update global totals
+        nonRebasingSupply = nonRebasingSupply.add(balanceOf(_account));
```

Comment:
Maybe you could explain here why you are using the updated `balanceOf` here, contrary to `_creditBalance` which you delibrately store before state changes are being made.